gnovm/pkg/gnolang/preprocess.go

```diff
@@ -2815,6 +2815,184 @@ func findUndefined(store Store, last BlockNode, x Expr) (un Name) {
 	return findUndefined2(store, last, x, nil)
 }
 
+// finds the next undefined identifier and returns it if it is global
+func findUndefined2SkipLocals(store Store, last BlockNode, x Expr, t Type, st *SymbolTable) Name {
+	name := findUndefined2(store, last, x, t)
+
+	if name != "" {
+		pkg := packageOf(last)
+
+		if _, _, ok := pkg.FileSet.GetDeclForSafe(name); !ok {
+			// skip it if it's a local identifier
+			return ""
+		}
+
+		isLocal := st.IdentifierExists(string(name))
+
+		if isLocal {
+			return ""
+		}
+	}
+
+	return name
+}
+
+func findUndefinedStmt(store Store, last BlockNode, stmt Stmt, t Type, st *SymbolTable) Name {
+	switch s := stmt.(type) {
+	case *BranchStmt:
+	case *ValueDecl:
+		for _, rh := range s.Values {
+			un := findUndefined2SkipLocals(store, last, rh, t, st)
+
+			if un != "" {
+				return un
+			}
+		}
+
+		for _, rh := range s.NameExprs {
+			st.AddIdentifier(string(rh.Name))
+		}
+	case *DeclStmt:
+		for _, rh := range s.Body {
+			un := findUndefinedStmt(store, last, rh, t, st)
+
+			if un != "" {
+				return un
+			}
+		}
+	case *IncDecStmt:
+		un := findUndefined2SkipLocals(store, last, s.X, t, st)
+
+		if un != "" {
+			return un
+		}
+	case *PanicStmt:
+		un := findUndefined2SkipLocals(store, last, s.Exception, t, st)
+
+		if un != "" {
+			return un
+		}
+	case *BlockStmt:
+		for _, rh := range s.Body {
+			un := findUndefinedStmt(store, last, rh, t, st)
+
+			if un != "" {
+				return un
+			}
+		}
+	case *DeferStmt:
+		un := findUndefined2SkipLocals(store, last, s.Call.Func, t, st)
+
+		if un != "" {
+			return un
+		}
+
+		for _, rh := range s.Call.Args {
+			un = findUndefined2SkipLocals(store, last, rh, t, st)
+
+			if un != "" {
+				return un
+			}
+		}
+	case *SwitchStmt:
+		un := findUndefined2SkipLocals(store, last, s.X, t, st)
+		if un != "" {
+			return un
+		}
+
+		un = findUndefinedStmt(store, last, s.Init, t, st)
+		if un != "" {
+			return un
+		}
+
+		for _, b := range s.Clauses {
+			b := b
+			un = findUndefinedStmt(store, last, &b, t, st)
+
+			if un != "" {
+				return un
+			}
+		}
+	case *SwitchClauseStmt:
+		for _, rh := range s.Cases {
+			un := findUndefined2SkipLocals(store, last, rh, t, st)
+
+			if un != "" {
+				return un
+			}
+		}
+
+		for _, b := range s.Body {
+			un := findUndefinedStmt(store, last, b, t, st)
+
+			if un != "" {
+				return un
+			}
+		}
+
+	case *ExprStmt:
+		return findUndefined2SkipLocals(store, last, s.X, t, st)
+	case *AssignStmt:
+		for i, rh := range s.Rhs {
+			if s.Op == DEFINE {
+				st.AddIdentifier(string(s.Lhs[i].(*NameExpr).Name))
+			}
+			un := findUndefined2SkipLocals(store, last, rh, t, st)
+
+			if un != "" {
+				return un
+			}
+		}
+	case *IfStmt:
+		un := findUndefined2SkipLocals(store, last, s.Cond, t, st)
+		if un != "" {
+			return un
+		}
+
+		un = findUndefinedStmt(store, last, &s.Else, t, st)
+		if un != "" {
+			return un
+		}
+
+		un = findUndefinedStmt(store, last, &s.Then, t, st)
+		if un != "" {
+			return un
+		}
```

Comment:
we can't push the same st in; all blocks should be creating their own scope.
"the way" is to use the BlockNode itself (here passed in as 'last'), rather than creating a parallel representation of chained/nested scopes. That is, the BlockNodes themselves should have initialized StaticBlocks, where name declarations can be stashed (and is read from through last in findUndefined when encounter a NameExpr). 

The problem is that al this predefine/findUndefined is supposed to happen before Predefine()--except for recursivity--but Predefine() is what currently initializes the BlockNode's StaticBlock. 

The solution generally should be to set up the StaticBlock during findUndefined*, as well as keep the existing initialization (but made idempotent if not already) in pushInitBlock/pushRealBlock, and to pass in the appropriate 'last' which depends on whether we are entering a new BlockNode or not. 

NOTE: during this findUndefined* traversal, probably should do the same thing as pushRealBlock, otherwise value paths within if/switch bodies will break later.

NOTE: see usage of `last2.Predefine`; you don't need the type nor (static) value for a predefine on a BlockNode; it's just reserving and registering the name on the block node's static block. The static type will be actually defined later during Preprocess. So just as tryPredefine() already calls BlockNode.Predefine(), so too should findUndefinedStmt call BlockNode.Predefine()

NOTE: I don't actually know whether pushInitBlock/pushRealBlock still would need to do any initialization after the above change, but it's probably safe to just make idempotent. Perhaps, if preprocessing changes the AST it matters.

NOTE: After initialization, a BlockNode links to the parent via BlockNode.StaticBlock.Block.Parent.Source. (see implementation of StaticBlock.InitStaticBlock and StaticBlock.GetParentNode) The parent does not separately link to child BlockNodes (besides the usual AST fields, such as IfStmt.Then). Then, .GetValueRef() can be used to check whether a name was (pre)defined or not.

Reply:
psuedocode where this becomes a problem
```go
var other = func() {
    if true {
        something := 2
        print(something) // 2
    } else {
        print(something) // a string, but single shared 'st' masks the outer/global reference. 
    }
}
var something = "a string"
```

Reply:
Working on a draft PR to solve this; https://github.com/gnolang/gno/pull/2418/files